In [1]:
# Add the Pandas dependecy
import pandas as pd
import os

In [2]:
# Files to load
school_data_to_load = os.path.join("Resources", "schools_complete.csv")
student_data_to_load = os.path.join("Resources", "students_complete.csv")

school_data_df = pd.read_csv(school_data_to_load)
student_data_df = pd.read_csv(student_data_to_load)

In [3]:
school_data_df.head()

School ID            school_name      type  size   budget
0          0      Huang High School  District  2917  1910635
1          1   Figueroa High School  District  2949  1884411
2          2    Shelton High School   Charter  1761  1056600
3          3  Hernandez High School  District  4635  3022020
4          4    Griffin High School   Charter  1468   917500

In [4]:
student_data_df.head()

Student ID       student_name gender grade        school_name  \
0           0       Paul Bradley      M   9th  Huang High School   
1           1       Victor Smith      M  12th  Huang High School   
2           2    Kevin Rodriguez      M  12th  Huang High School   
3           3  Dr. Richard Scott      M  12th  Huang High School   
4           4         Bonnie Ray      F   9th  Huang High School   

   reading_score  math_score  
0             66          79  
1             94          61  
2             90          60  
3             67          58  
4             97          84

In [5]:
# Determine if there are any missing values in the school data.
school_data_df.count()

School ID      15
school_name    15
type           15
size           15
budget         15
dtype: int64

In [6]:
# Determine if there are any missing values in the student data.
student_data_df.count()

Student ID       39170
student_name     39170
gender           39170
grade            39170
school_name      39170
reading_score    39170
math_score       39170
dtype: int64

There are no missing values in any of the columns because the output shows 39,170 rows for each column in student_data_df. By feault, "null" values are not counted, so you can often quickly identify which columns have missing data.

In [7]:
# Determine if there are any missing values in the school data.
school_data_df.isnull()

School ID  school_name   type   size  budget
0       False        False  False  False   False
1       False        False  False  False   False
2       False        False  False  False   False
3       False        False  False  False   False
4       False        False  False  False   False
5       False        False  False  False   False
6       False        False  False  False   False
7       False        False  False  False   False
8       False        False  False  False   False
9       False        False  False  False   False
10      False        False  False  False   False
11      False        False  False  False   False
12      False        False  False  False   False
13      False        False  False  False   False
14      False        False  False  False   False

We see that every row that is not empty is given the Boolean value "False", which tells us that there are no missing values. This is another method to determine empty rows.

In [8]:
student_data_df.isnull()

Student ID  student_name  gender  grade  school_name  reading_score  \
0           False         False   False  False        False          False   
1           False         False   False  False        False          False   
2           False         False   False  False        False          False   
3           False         False   False  False        False          False   
4           False         False   False  False        False          False   
...           ...           ...     ...    ...          ...            ...   
39165       False         False   False  False        False          False   
39166       False         False   False  False        False          False   
39167       False         False   False  False        False          False   
39168       False         False   False  False        False          False   
39169       False         False   False  False        False          False   

       math_score  
0           False  
1           False  
2           False  
3           False  
4           False  
...           ...  
39165       False  
39166       False  
39167       False  
39168       False  
39169       False  

[39170 rows x 7 columns]

In [9]:
# To get the total number of empty rows, or rows that are "True", we can use the Pandas ".sum()" method
# after the ".isnull()" method.
student_data_df.isnull().sum()

Student ID       0
student_name     0
gender           0
grade            0
school_name      0
reading_score    0
math_score       0
dtype: int64

In [10]:
# The ".notnull()" method returns the opposite boolean value of an ".isnull()" method
student_data_df.notnull()

Student ID  student_name  gender  grade  school_name  reading_score  \
0            True          True    True   True         True           True   
1            True          True    True   True         True           True   
2            True          True    True   True         True           True   
3            True          True    True   True         True           True   
4            True          True    True   True         True           True   
...           ...           ...     ...    ...          ...            ...   
39165        True          True    True   True         True           True   
39166        True          True    True   True         True           True   
39167        True          True    True   True         True           True   
39168        True          True    True   True         True           True   
39169        True          True    True   True         True           True   

       math_score  
0            True  
1            True  
2            True  
3            True  
4            True  
...           ...  
39165        True  
39166        True  
39167        True  
39168        True  
39169        True  

[39170 rows x 7 columns]

In [11]:
student_data_df.notnull().sum()

Student ID       39170
student_name     39170
gender           39170
grade            39170
school_name      39170
reading_score    39170
math_score       39170
dtype: int64

In [12]:
# Determine the data types included in the school DataFrame
school_data_df.dtypes

School ID       int64
school_name    object
type           object
size            int64
budget          int64
dtype: object

In [13]:
# Determine the data types included in the student DataFrame
student_data_df.dtypes

Student ID        int64
student_name     object
gender           object
grade            object
school_name      object
reading_score     int64
math_score        int64
dtype: object

See cleaning_student_names.ipynb to see the process used to clean student name data

In [14]:
# Add each prefix and suffix to remove to a list.
prefixes_suffixes = ["Dr. ", "Mr. ","Ms. ", "Mrs. ", "Miss ", " MD", " DDS", " DVM", " PhD"]

# Iterate through the words in the "prefixes_suffixes" list and replace them with an empty space, "".
for word in prefixes_suffixes:
    student_data_df["student_name"] = student_data_df["student_name"].str.replace(word,"")

In [15]:
student_data_df

Student ID     student_name gender grade         school_name  \
0               0     Paul Bradley      M   9th   Huang High School   
1               1     Victor Smith      M  12th   Huang High School   
2               2  Kevin Rodriguez      M  12th   Huang High School   
3               3    Richard Scott      M  12th   Huang High School   
4               4       Bonnie Ray      F   9th   Huang High School   
...           ...              ...    ...   ...                 ...   
39165       39165     Donna Howard      F  12th  Thomas High School   
39166       39166        Dawn Bell      F  10th  Thomas High School   
39167       39167   Rebecca Tanner      F   9th  Thomas High School   
39168       39168     Desiree Kidd      F  10th  Thomas High School   
39169       39169  Carolyn Jackson      F  11th  Thomas High School   

       reading_score  math_score  
0                 66          79  
1                 94          61  
2                 90          60  
3                 67          58  
4                 97          84  
...              ...         ...  
39165             99          90  
39166             95          70  
39167             73          84  
39168             99          90  
39169             95          75  

[39170 rows x 7 columns]

Now that we have cleaned the dataframes, its best practice to create a new dataframe to do calculations. This way, the original data is not affected. We will merge the two data frames on the shared column "school_name". There may be cases in which you want to merge on a column that has similar information in two seperate data frame, however they're name differently. Its goot practice to rename the columns so that they match. This will prevent duplicate columns are merging issues.

In [16]:
# Combine the data into a single dataset.
school_data_complete_df = pd.merge(student_data_df, school_data_df, on=["school_name", "school_name"])
school_data_complete_df.head()

Student ID     student_name gender grade        school_name  reading_score  \
0           0     Paul Bradley      M   9th  Huang High School             66   
1           1     Victor Smith      M  12th  Huang High School             94   
2           2  Kevin Rodriguez      M  12th  Huang High School             90   
3           3    Richard Scott      M  12th  Huang High School             67   
4           4       Bonnie Ray      F   9th  Huang High School             97   

   math_score  School ID      type  size   budget  
0          79          0  District  2917  1910635  
1          61          0  District  2917  1910635  
2          60          0  District  2917  1910635  
3          58          0  District  2917  1910635  
4          84          0  District  2917  1910635

In [18]:
# Get the total number of students
student_count = school_data_complete_df["Student ID"].count()
student_count

39170

In [23]:
# Calculate the total number of schools in school_data_df
school_count = school_data_df["school_name"].count()
print("There are " + str(school_count) + " number of schools in school_data_df")

# Calculate the total number of schools in school_data_complete_df
school_count2 = school_data_complete_df["school_name"].unique()
print("There are " + str(len(school_count2)) + " number of schools in school_data_complete_df")

There are 15 number of schools in school_data_df
There are 15 number of schools in school_data_complete_df


In [24]:
# Calculate total budget
total_budget = school_data_df["budget"].sum()
print("The total budget in the district is $" + str(total_budget))

The total budget in the district is $24649428


In [25]:
# Find average reading score
average_reading_score = school_data_complete_df["reading_score"].mean()
print("The average reading score is " + str(average_reading_score))

The average reading score is 81.87784018381414


In [26]:
# Find the average math score
average_math_score = school_data_complete_df["math_score"].mean()
print("The average math score is " + str(average_math_score))

The average math score is 78.98537145774827


We determine a passing score was 70 for both the reading and math assessments. To determine the percentage of students that passes both, we will find the total count of students who recieved >= 70, and divide that by the total number of students.

In [31]:
# Calculate the number of students passed the math assessment
passing_math = school_data_complete_df[school_data_complete_df["math_score"] >= 70]
passing_math_count = passing_math["student_name"].count()
passing_math_percent = passing_math_count/student_count * 100
print("A total of %" + str(passing_math_percent) + " passed the math exam")

A total of %74.9808526933878 passed the math exam


In [37]:
# Calculate the number of students who passed the reading assessment
passing_reading = school_data_complete_df[school_data_complete_df["reading_score"] >= 70]
passing_reading_count = passing_reading["student_name"].count()
passing_reading_percent = passing_reading_count/student_count * 100
print("A total of %" + str(passing_reading_percent) + " passed the reading exam")

A total of %85.80546336482001 passed the reading exam


In [39]:
# Calculate the Overall Passing Percentage
passing_math_reading = school_data_complete_df[(school_data_complete_df["math_score"] >= 70) & (school_data_complete_df["reading_score"] >= 70)]
passing_math_reading_count = passing_math_reading["student_name"].count()
passing_math_reading_percent = passing_math_reading_count/student_count * 100
print("A total of %" + str(passing_math_reading_percent) + " passed both exams")

A total of %65.17232575950983 passed both exams


In [42]:
# Adding a list of values with keys to create a new DataFrame.
district_summary_df = pd.DataFrame(
          [{"Total Schools": school_count,
          "Total Students": student_count,
          "Total Budget": total_budget,
          "Average Math Score": average_math_score,
          "Average Reading Score": average_reading_score,
          "% Passing Math": passing_math_percent,
         "% Passing Reading": passing_reading_percent,
        "% Overall Passing": passing_math_reading_percent}])
district_summary_df

Total Schools  Total Students  Total Budget  Average Math Score  \
0             15           39170      24649428           78.985371   

   Average Reading Score  % Passing Math  % Passing Reading  % Overall Passing  
0               81.87784       74.980853          85.805463          65.172326

The Pandas ".map()" function is used for substituting each value in a Series with another value. Where the new value is generated from a function, a dictionary, or a series. This will be useful when formatting district_summary_df.

series.map("current_value_1" : "new_value_1",  "current_value_2" : "new_value_2", etc)

The Python ".format()" function is used to format a value to a specific format, such as one or two decimal places, or adding a thousands seperator.

"{value:format specification}".format(value)

In [43]:
# Format "Total Budget" to have the comma for a thousands separator, a decimal separator, and a "$".

district_summary_df["Total Budget"] = district_summary_df["Total Budget"].map("${:,.2f}".format)

district_summary_df["Total Budget"]

0    $24,649,428.00
Name: Total Budget, dtype: object

In [45]:
# Format the columns.
district_summary_df["Average Math Score"] = district_summary_df["Average Math Score"].map("{:.1f}".format)

district_summary_df["Average Reading Score"] = district_summary_df["Average Reading Score"].map("{:.1f}".format)

district_summary_df["% Passing Math"] = district_summary_df["% Passing Math"].map("{:.0f}".format)

district_summary_df["% Passing Reading"] = district_summary_df["% Passing Reading"].map("{:.0f}".format)

district_summary_df["% Overall Passing"] = district_summary_df["% Overall Passing"].map("{:.0f}".format)

In [46]:
# Reorder the columns in the order you want them to appear.
new_column_order = ["Total Schools", "Total Students", "Total Budget","Average Math Score", "Average Reading Score", "% Passing Math", "% Passing Reading", "% Overall Passing"]

# Assign district summary df the new column order.
district_summary_df = district_summary_df[new_column_order]
district_summary_df

Total Schools  Total Students    Total Budget Average Math Score  \
0             15           39170  $24,649,428.00               79.0   

  Average Reading Score % Passing Math % Passing Reading % Overall Passing  
0                  81.9             75                86                65